### DISCUSSION TITLE
Microstructure Generation in Neper using EBSD Data

### DISCUSSION THREAD

**Discussion by dikshanshs** [neperfepx/neper#258] [Q&A] [answered]
Hi @rquey Sir,
I have EBSD Data and, I want to use this data to generate microstructure using Neper. So that, I can further work on this microstructure. Please tell me how to use this data?
Note: Data I have is in .ctf and .ang files.

Edited: I have extracted voxel data and orientation.
@preethamalluri please help me out with this

**Comment by rquey** [comment, marked as answer]
To load the map in Neper, the first thing is to write the EBSD data as a tesr file. See the documentation for specifications on the tesr file. Then, you will be able to "view" the data (in -V), generate a tess file from it (in -T), mesh it (in -M), etc.

**Reply by dikshanshs** [reply]
@rquey sir,
                  I tried copying binary data (from VTK) for given EBSD data and also made TESR file but it didn't work. Please share the procedure of creating a TESR file using EBSD data, and Neper syntax used. 
EBSD Data : [dis - Copy.zip](https://github.com/rquey/neper/files/6803203/dis.-.Copy.zip)

**Reply by rquey** [reply]
See the documentation for a description of the tesr file. You can also generate one and look into it. Here is how you can generate a EBSD-alike tesr:

```
$ neper -T -n 10 -dim 2 -tesrsize 40 -for tesr -oridistrib "normal(1)"
$ neper -V n10-id1.tesr -datavoxcol ori -datavoxcolscheme ipf -print img
```
![image](https://user-images.githubusercontent.com/2659376/125341742-4471d180-e354-11eb-9fca-a3ead34d7fdc.png)

**Reply by dikshanshs** [reply]
@rquey Sir, 
 If possible please share a tesr file that is generated using EBSD data or Voxel data (XY & orientation).
as I am facing a problem in interpreting the tesr file structure.

**Reply by rquey** [reply]
`$ neper -T -n 10 -dim 2 -tesrsize 40 -for tesr -oridistrib "normal(1)"` gives you the tesr file. Here is a minimal version of that file: [n10-id1.tesr.txt](https://github.com/rquey/neper/files/6821783/n10-id1.tesr.txt).

Notably, the file must contain the pixel orientations (`**oridata`) and the pixel grain ids (`**data`).

**Reply by venki9494** [reply]
I am a use trying to convert EBSD data.i am using the above code however i an unable to get the n10-id1.tesr.text file you are talking about containing **oridata and **data.instead i get the attached file.Quite confusing it is!!

**Comment by Ming-is** [comment]
> 嗨@rquey先生， 我有 EBSD 数据，我想使用这些数据使用 Neper 生成微观结构。这样，我就可以进一步研究这种微观结构。请告诉我如何使用这些数据？ 注意：我拥有的数据在 .ctf 和 .ang 文件中。
> 
> 编辑：我已经提取了体素数据和方向。 @preethamalluri请帮我解决这个问题

dear，Have you solved this problem? I also have EBSD ctf .file and ang .file, but I don't know how to use these data to become neper usable files.Looking forward to your reply, thank you

**Reply by dikshanshs** [reply]
@Ming-is,
I used the following steps:
![image](https://user-images.githubusercontent.com/85054451/141724076-c486e744-504a-4797-9019-0f46b4ba67d8.png)

**Reply by Ming-is** [reply]
Regarding the first step, can you recommend the ‘’pipeline ‘’ to be used in dream3d ？

**Reply by dikshanshs** [reply]
INL Export

**Reply by Ming-is** [reply]
I'm still confused about some parameter settings in pipeline, such as' feature Ids'. Can you share your complete pipeline bookmarks file?

And can you send me a copy of the TESR file you created, so that I can modify it against my own file.
thank you very much。

**Reply by BK-456** [reply]
I am facing issue regarding the creation of .tess file from the EBSD .ctf data file. Can anybody please share the procedure and Neper syntax for the same.

**Reply by Anmol-Bakshi** [reply]
For edge:
Describe edge from top vertex to bottom vertex or, from left vertex to right vertex.
If both are applicable at the same time, then edge described by top vertex to bottom vertex definition dominates

For face:
vertex: starting from any vertex write in clockwise direction
edge: follow the sequence of vertex description to write edge, i.e., in clockwise direction. But suppose previously, edge 7 has been defined from vertex 4 to vertex 5 and now for maintaining the clockwise direction, edge 7 has to be described from vertex 5 to vertex 4, then write edge 7 as '-7'.
